<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200> 

# Assignment:

- Replicate the lesson code.

 - This means that if you haven't followed along already, type out the things that we did in class. Forcing your fingers to hit each key will help you internalize the syntax of what we're doing. Make sure you understand each line of code that you're writing, google things that you don't fully understand.
 - [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)
- Convert the `term` column from string to integer.
- Make a column named `loan_status_is_great`. It should contain the integer 1 if `loan_status` is "Current" or "Fully Paid." Else it should contain the integer 0.
- Make `last_pymnt_d_month` and `last_pymnt_d_year` columns.

# BEGIN ASSIGNMENT REPLICATE

In [1]:
##### Begin Working Here #####
# Import necessary packages and set display options for Dataset.

import pandas as pd
import numpy as np

pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 1461)

In [2]:
# Fetch LendingClub Data via BASH command
!wget 'https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip'

# Unzip dataset to CSV
!unzip LoanStats_2018Q4.csv.zip

--2020-09-05 04:07:34--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 35.161.89.82, 54.148.13.215, 54.244.115.45
Connecting to resources.lendingclub.com (resources.lendingclub.com)|35.161.89.82|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip.3’

LoanStats_2018Q4.cs     [               <=>  ]  22.28M  6.22MB/s    in 3.7s    

2020-09-05 04:07:38 (6.01 MB/s) - ‘LoanStats_2018Q4.csv.zip.3’ saved [23360898]

Archive:  LoanStats_2018Q4.csv.zip
replace LoanStats_2018Q4.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


In [3]:
# Read in the unzipped CSV file to Pandas
df = pd.read_csv('LoanStats_2018Q4.csv');

# Examine the head to insure proper formating
df.head(10)

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0,1,2,3,4,7,13,18,19,24,25,27,28,29,30,31,32,34,36,37,38,39,40,41,42,43,44,46,49,50,51,53,54,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,113,114,115,116,117,118,119,120,121,126,127,131,132,134,135,136,141,142,143) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id  member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title                        emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url desc purpose            title                   zip_code addr_state dti   delinq_2yrs earliest_cr_line inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt      total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason    hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
NaN NaN       27975     27975       27975            36 months  14.47%  962.52      C     C2        Conductor                        10+ years  MORTGAGE       180000     Not Verified        Dec-2018 Fully Paid  n          NaN NaN  credit_card        Credit card refinancing 117xx    NY         11.47 0           Jul-1995         0              39                     NaN                    10       0       29711     66.8%      19        w                   0.00      0.00          31804.5298492634 31804.53        27975.00        3829.53       0.0                0.0        0.0                     Jan-2020     20288.02        NaN          Mar-2020           0                          NaN                         1           Individual       NaN              NaN       NaN                       0              0            206072      0           1           0           1           17                 15676        63      0           1           10861      65       44500            1      4           2            2                    20607       7416           74.9    0                        0           17                 281                  20                    17             5        20                   39                       1                     39                             0                     5              7               5           6         1         8             13            7                   10       0                0            0                  0             

##### Looks like the first line is causing some issues with NaN values! Let's clean this up by skipping that first row with the URL!

In [4]:
# Assign skiprow to ignore the URL
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)

df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN    27975.0      27975.0          27975.0   36 months   
1  NaN        NaN    10000.0      10000.0          10000.0   60 months   
2  NaN        NaN    28000.0      28000.0          28000.0   60 months   
3  NaN        NaN    16000.0      16000.0          16000.0   60 months   
4  NaN        NaN     7500.0       7500.0           7500.0   36 months   

  int_rate  installment grade sub_grade        emp_title emp_length  \
0   14.47%       962.52     C        C2       Conductor   10+ years   
1   12.98%       227.43     B        B5          Printer    9 years   
2   13.56%       645.15     C        C1  Project Manager  10+ years   
3   13.56%       368.66     C        C1             LMSW  10+ years   
4   10.72%       244.55     B        B2  Program Manager    2 years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0       MORTGAGE    180000.0        Not Verified  Dec-2018   Fully Paid   
1           RENT     60000.0        Not Verified  Dec-2018      Current   
2       MORTGAGE    128500.0     Source Verified  Dec-2018      Current   
3           RENT     46000.0        Not Verified  Dec-2018      Current   
4           RENT     84000.0        Not Verified  Dec-2018  Charged Off   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    117xx   
1          n  NaN   NaN         credit_card  Credit card refinancing    660xx   
2          n  NaN   NaN         credit_card  Credit card refinancing    760xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    787xx   
4          n  NaN   NaN  debt_consolidation       Debt consolidation    600xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  11.47          0.0         Jul-1995             0.0   
1         KS  14.90          0.0         May-2007             0.0   
2         TX  27.35          1.0         Jul-1987             0.0   
3         TX  11.09          1.0         Oct-1997             1.0   
4         IL   3.86          0.0         May-2005             1.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                    39.0                     NaN      10.0      0.0   
1                     NaN                   112.0       7.0      1.0   
2                    24.0                     NaN      16.0      0.0   
3                    11.0                     NaN       9.0      0.0   
4                     NaN                   114.0       9.0      1.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0    29711.0      66.8%       19.0                   w       0.00   
1    10677.0      54.2%       12.0                   w    7352.95   
2    55206.0        75%       31.0                   w   20667.91   
3    18946.0      32.1%       21.0                   w   11503.84   
4     2200.0      12.6%       17.0                   w       0.00   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00  31804.529849         31804.53         27975.00   
1        7352.95   4487.310000          4487.31          2647.05   
2       20667.91  12871.360000         12871.36          7332.09   
3       11503.84   7570.880000          7570.88          4496.16   
4           0.00   2829.930000          2829.93          1465.61   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        3829.53                 0.0        0.00                   0.0000   
1        1840.26                 0.0        0.00                   0.0000   
2        5539.27                 0.0        0.00                   0.0000   
3        3074.72                 0.0        0.00                   0.0000   
4         484.09                 0.0      880.23                 158.4414   

  last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pu

##### Looks better! Now we can take a look at these NaN values!

In [5]:
df.isnull().sum().sort_values()

inq_fi                                             2
mo_sin_old_rev_tl_op                               2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
total_cu_tl                                        2
total_rev_hi_lim                                   2
open_rv_24m                                        2
open_rv_12m                                        2
total_bal_il                                       2
open_il_24m                                        2
open_il_12m                                        2
open_act_il                                        2
open_acc_6m                                        2
tot_cur_bal                                        2
tot_coll_amt                                       2
acc_now_delinq                                     2
application_type                              

##### Well that's odd....we have a TON of 2's...let's take a look at the tail and see what's going on.

In [6]:
df.tail()

id  member_id  \
128409                                               NaN        NaN   
128410                                               NaN        NaN   
128411                                               NaN        NaN   
128412  Total amount funded in policy code 1: 2050909275        NaN   
128413   Total amount funded in policy code 2: 820109297        NaN   

        loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
128409     5000.0       5000.0           5000.0   36 months   13.56%   
128410    10000.0      10000.0           9750.0   36 months   11.06%   
128411    10000.0      10000.0          10000.0   36 months   16.91%   
128412        NaN          NaN              NaN         NaN      NaN   
128413        NaN          NaN              NaN         NaN      NaN   

        installment grade sub_grade             emp_title emp_length  \
128409       169.83     C        C1          Payoff Clerk  10+ years   
128410       327.68     B        B3                   NaN        NaN   
128411       356.08     C        C5  Key Accounts Manager    2 years   
128412          NaN   NaN       NaN                   NaN        NaN   
128413          NaN   NaN       NaN                   NaN        NaN   

       home_ownership  annual_inc verification_status   issue_d loan_status  \
128409       MORTGAGE     35360.0        Not Verified  Oct-2018     Current   
128410           RENT     44400.0     Source Verified  Oct-2018     Current   
128411           RENT     80000.0        Not Verified  Oct-2018     Current   
128412            NaN         NaN                 NaN       NaN         NaN   
128413            NaN         NaN                 NaN       NaN         NaN   

       pymnt_plan  url  desc             purpose                    title  \
128409          n  NaN   NaN  debt_consolidation       Debt consolidation   
128410          n  NaN   NaN         credit_card  Credit card refinancing   
128411          n  NaN   NaN               other                    Other   
128412        NaN  NaN   NaN                 NaN                      NaN   
128413        NaN  NaN   NaN                 NaN                      NaN   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128409    381xx         TN  11.30          1.0         Jun-2006   
128410    980xx         WA  11.78          0.0         Oct-2008   
128411    021xx         MA  17.72          1.0         Sep-2006   
128412      NaN        NaN    NaN          NaN              NaN   
128413      NaN        NaN    NaN          NaN              NaN   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128409             0.0                    21.0                     NaN   
128410             2.0                    40.0                     NaN   
128411             0.0                    14.0                     NaN   
128412             NaN                     NaN                     NaN   
128413             NaN                     NaN                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128409       9.0      0.0     2597.0      27.3%       15.0   
128410      15.0      0.0     6269.0      13.1%       25.0   
128411      17.0      0.0     1942.0      30.8%       31.0   
128412       NaN      NaN        NaN        NaN        NaN   
128413       NaN      NaN        NaN        NaN        NaN   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128409                   f    2187.38        2187.38      3732.49   
128410                   f    4285.08        4177.96      7193.60   
128411                   w    4495.59        4495.59      7824.37   
128412                 NaN        NaN            NaN          NaN   
128413                 NaN        NaN            NaN          NaN   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
128409          3732.49          2812.62         919.87                 0.0   
128410          7013.76      

##### Looks like we have some extra values in the last two rows as well. Let's clean this up!

In [7]:
# Reload dataframe so that the first line is skipped and the last two lines are skipped.
# We will also pass the 'engine=python' argument to avoid the C error.
df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2, engine='python')

df.tail(5)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
128407 NaN        NaN      23000        23000          23000.0   36 months   
128408 NaN        NaN      10000        10000          10000.0   36 months   
128409 NaN        NaN       5000         5000           5000.0   36 months   
128410 NaN        NaN      10000        10000           9750.0   36 months   
128411 NaN        NaN      10000        10000          10000.0   36 months   

       int_rate  installment grade sub_grade             emp_title emp_length  \
128407   15.02%       797.53     C        C3        Tax Consultant  10+ years   
128408   15.02%       346.76     C        C3        security guard    5 years   
128409   13.56%       169.83     C        C1          Payoff Clerk  10+ years   
128410   11.06%       327.68     B        B3                   NaN        NaN   
128411   16.91%       356.08     C        C5  Key Accounts Manager    2 years   

       home_ownership  annual_inc verification_status   issue_d  loan_status  \
128407       MORTGAGE     75000.0     Source Verified  Oct-2018  Charged Off   
128408       MORTGAGE     38000.0        Not Verified  Oct-2018      Current   
128409       MORTGAGE     35360.0        Not Verified  Oct-2018      Current   
128410           RENT     44400.0     Source Verified  Oct-2018      Current   
128411           RENT     80000.0        Not Verified  Oct-2018      Current   

       pymnt_plan  url  desc             purpose                    title  \
128407          n  NaN   NaN  debt_consolidation       Debt consolidation   
128408          n  NaN   NaN  debt_consolidation       Debt consolidation   
128409          n  NaN   NaN  debt_consolidation       Debt consolidation   
128410          n  NaN   NaN         credit_card  Credit card refinancing   
128411          n  NaN   NaN               other                    Other   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128407    352xx         AL  20.95            1         Aug-1985   
128408    443xx         OH  13.16            3         Jul-1982   
128409    381xx         TN  11.30            1         Jun-2006   
128410    980xx         WA  11.78            0         Oct-2008   
128411    021xx         MA  17.72            1         Sep-2006   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128407               2                    22.0                     NaN   
128408               0                     6.0                     NaN   
128409               0                    21.0                     NaN   
128410               2                    40.0                     NaN   
128411               0                    14.0                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128407        12        0      22465      43.6%         28   
128408        11        0       5634      37.1%         16   
128409         9        0       2597      27.3%         15   
128410        15        0       6269      13.1%         25   
128411        17        0       1942      30.8%         31   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128407                   w       0.00           0.00      1547.08   
128408                   w    4427.45        4427.45      7620.38   
128409                   f    2187.38        2187.38      3732.49   
128410                   f    4285.08        4177.96      7193.60   
128411                   w    4495.59        4495.59      7824.37   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
128407          1547.08          1025.67         521.41                 0.0   
128408          7620.38          5572.55        2047.83                 0.0   
128409          3732.49          2812.62         919.87                 0.0   
128410          7013.76          5714.92        1478.68                 0.0   
128411          7824.37          5504.41        2319.96                 0.0  

##### Now that we have our table formatted properly we can get to work! Let's start by dropping any "unnecessary" values from our dataset that are equal to NaN!

In [8]:
df = df.drop(['id', 'member_id', 'url', 'desc', 'settlement_status', 
              'settlement_amount', 'settlement_date', 'settlement_term', 
              'settlement_percentage', 'debt_settlement_flag_date'], axis=1)

In [9]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      27975        27975          27975.0   36 months   14.47%       962.52   
1      10000        10000          10000.0   60 months   12.98%       227.43   
2      28000        28000          28000.0   60 months   13.56%       645.15   
3      16000        16000          16000.0   60 months   13.56%       368.66   
4       7500         7500           7500.0   36 months   10.72%       244.55   

  grade sub_grade        emp_title emp_length home_ownership  annual_inc  \
0     C        C2       Conductor   10+ years       MORTGAGE    180000.0   
1     B        B5          Printer    9 years           RENT     60000.0   
2     C        C1  Project Manager  10+ years       MORTGAGE    128500.0   
3     C        C1             LMSW  10+ years           RENT     46000.0   
4     B        B2  Program Manager    2 years           RENT     84000.0   

  verification_status   issue_d  loan_status pymnt_plan             purpose  \
0        Not Verified  Dec-2018   Fully Paid          n         credit_card   
1        Not Verified  Dec-2018      Current          n         credit_card   
2     Source Verified  Dec-2018      Current          n         credit_card   
3        Not Verified  Dec-2018      Current          n         credit_card   
4        Not Verified  Dec-2018  Charged Off          n  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    117xx         NY  11.47            0   
1  Credit card refinancing    660xx         KS  14.90            0   
2  Credit card refinancing    760xx         TX  27.35            1   
3  Credit card refinancing    787xx         TX  11.09            1   
4       Debt consolidation    600xx         IL   3.86            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Jul-1995               0                    39.0   
1         May-2007               0                     NaN   
2         Jul-1987               0                    24.0   
3         Oct-1997               1                    11.0   
4         May-2005               1                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        10        0      29711      66.8%         19   
1                   112.0         7        1      10677      54.2%         12   
2                     NaN        16        0      55206        75%         31   
3                     NaN         9        0      18946      32.1%         21   
4                   114.0         9        1       2200      12.6%         17   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  31804.529849   
1                   w    7352.95        7352.95   4487.310000   
2                   w   20667.91       20667.91  12871.360000   
3                   w   11503.84       11503.84   7570.880000   
4                   w       0.00           0.00   2829.930000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         31804.53         27975.00        3829.53                 0.0   
1          4487.31          2647.05        1840.26                 0.0   
2         12871.36          7332.09        5539.27                 0.0   
3          7570.88          4496.16        3074.72                 0.0   
4          2829.93          1465.61         484.09                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0        0.00                   0.0000     Jan-2020         20288.02   
1        0.00                   0.0000     Aug-2020           227.43   
2        0.00                   0.0000     Aug-2020           645.15   
3        0.00                   0.0000     Aug-2020           368.66   
4      880.23                 158.4414     Sep-2019           244.55   

  next_pymnt_d last_credit_pull_d  colle

##### Now we'll take a look at the 'int_rate' column and see if we can clean it up. It shows a '%' sign now which indicates that this is a string value.

In [10]:
# Examine 'int_rate'
df['int_rate'].head(10)

0     14.47%
1     12.98%
2     13.56%
3     13.56%
4     10.72%
5     20.89%
6     26.31%
7     23.40%
8     19.92%
9     17.97%
Name: int_rate, dtype: object

##### Looks like we were right! Let's drop any whitespace and the '%' from this column!

In [11]:
# First we will define a function to make our lives easier
def intrate_to_float(cell_contents):
  return float(cell_contents.strip().strip('%'))

# Now we will use .apply to call the function on our column
df['int_rate'] = df['int_rate'].apply(intrate_to_float)

In [12]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      27975        27975          27975.0   36 months     14.47       962.52   
1      10000        10000          10000.0   60 months     12.98       227.43   
2      28000        28000          28000.0   60 months     13.56       645.15   
3      16000        16000          16000.0   60 months     13.56       368.66   
4       7500         7500           7500.0   36 months     10.72       244.55   

  grade sub_grade        emp_title emp_length home_ownership  annual_inc  \
0     C        C2       Conductor   10+ years       MORTGAGE    180000.0   
1     B        B5          Printer    9 years           RENT     60000.0   
2     C        C1  Project Manager  10+ years       MORTGAGE    128500.0   
3     C        C1             LMSW  10+ years           RENT     46000.0   
4     B        B2  Program Manager    2 years           RENT     84000.0   

  verification_status   issue_d  loan_status pymnt_plan             purpose  \
0        Not Verified  Dec-2018   Fully Paid          n         credit_card   
1        Not Verified  Dec-2018      Current          n         credit_card   
2     Source Verified  Dec-2018      Current          n         credit_card   
3        Not Verified  Dec-2018      Current          n         credit_card   
4        Not Verified  Dec-2018  Charged Off          n  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    117xx         NY  11.47            0   
1  Credit card refinancing    660xx         KS  14.90            0   
2  Credit card refinancing    760xx         TX  27.35            1   
3  Credit card refinancing    787xx         TX  11.09            1   
4       Debt consolidation    600xx         IL   3.86            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Jul-1995               0                    39.0   
1         May-2007               0                     NaN   
2         Jul-1987               0                    24.0   
3         Oct-1997               1                    11.0   
4         May-2005               1                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        10        0      29711      66.8%         19   
1                   112.0         7        1      10677      54.2%         12   
2                     NaN        16        0      55206        75%         31   
3                     NaN         9        0      18946      32.1%         21   
4                   114.0         9        1       2200      12.6%         17   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  31804.529849   
1                   w    7352.95        7352.95   4487.310000   
2                   w   20667.91       20667.91  12871.360000   
3                   w   11503.84       11503.84   7570.880000   
4                   w       0.00           0.00   2829.930000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         31804.53         27975.00        3829.53                 0.0   
1          4487.31          2647.05        1840.26                 0.0   
2         12871.36          7332.09        5539.27                 0.0   
3          7570.88          4496.16        3074.72                 0.0   
4          2829.93          1465.61         484.09                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0        0.00                   0.0000     Jan-2020         20288.02   
1        0.00                   0.0000     Aug-2020           227.43   
2        0.00                   0.0000     Aug-2020           645.15   
3        0.00                   0.0000     Aug-2020           368.66   
4      880.23                 158.4414     Sep-2019           244.55   

  next_pymnt_d last_credit_pull_d 

##### Looks good! But let's check with dtype to be sure.

In [13]:
# Call dtypes on 'int_rate' to ensure float
df['int_rate'].dtypes

dtype('float64')

##### Next we will take a look at the employment title column as according to 'isnull()' there was some NaN values

In [14]:
df['emp_title'].value_counts(dropna=False)[:10]

NaN                 20947
Teacher              2090
Manager              1773
Registered Nurse      952
Driver                924
RN                    726
Supervisor            697
Sales                 580
Project Manager       526
General Manager       523
Name: emp_title, dtype: int64

##### Holy cow! We have 20,947 NaN Values in Employment Title! Let's get that fixed.


In [15]:
# Define function for employment title cleaning
def clean_employment(emp_title):
  if isinstance(emp_title, str):
    return emp_title.title().strip()
  else:
    return "Unknown"

In [16]:
# Now we can apply our function to our original 'emp_title' column.
df['emp_title'] = df['emp_title'].apply(clean_employment)

In [17]:
# Examine 'emp_title' for cleanliness
df['emp_title'].value_counts(ascending=False)

Unknown                                     20947
Teacher                                      2557
Manager                                      2395
Registered Nurse                             1418
Driver                                       1258
                                            ...  
Assistant Vice President, Branch Manager        1
Assistant Director Of School Programs           1
Warehouseworker Ii                              1
Programming Supervisor                          1
Director Of Insurance                           1
Name: emp_title, Length: 34902, dtype: int64

##### Now we have our 'emp_title' column cleared of any NaN values! Now we'll work with the date columns next.

In [18]:
# List comprehension to print date columns
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

In [19]:
df['issue_d'][:5]

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
Name: issue_d, dtype: object

##### It looks like our 'issue_d' column is set as MMM-YYYY. (Three digit month, four digit year)

##### However, it is showing as a 'String' instead of 'Datetime'! Let's utilize pandas Datetime function to fix that!

In [20]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format = True)

df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                      float64
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

##### Now that we have our 'issue_d' displaying as Datetime correctly, we can extract the years and months for each loan and append new columns to our dataset.

In [21]:
df['issue_month'] = df['issue_d'].dt.month
df['issue_year'] = df['issue_d'].dt.year

##### Let's look at our last two columns to ensure they were added correctly!

In [22]:
# Call dtypes on the last two columns to ensure .dt works correctly
df.dtypes[-2:]

issue_month    int64
issue_year     int64
dtype: object

##### Perfect! Unfortunately there's only three months available as this is strictly for Q4. Let's take a look at 'earliest_cr_line' as that should be much more varied.

In [23]:
df['earliest_cr_line'].head()

0    Jul-1995
1    May-2007
2    Jul-1987
3    Oct-1997
4    May-2005
Name: earliest_cr_line, dtype: object

##### This looks much more varied on both terms! Yet if we look closer it's another String! Let's apply 'to_datetime' to this column and see what cool features we can build with it!

In [24]:
# Apply to_datetime on 'earliest_cr_line' to break string format.
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [25]:
# Verify datetime was successful
df['earliest_cr_line'][:10]

0   1995-07-01
1   2007-05-01
2   1987-07-01
3   1997-10-01
4   2005-05-01
5   2004-10-01
6   2010-09-01
7   2007-05-01
8   2005-04-01
9   2011-04-01
Name: earliest_cr_line, dtype: datetime64[ns]

##### Let's make a new column to determine the time difference between first line of credit and the LendingClub loan

In [26]:
# Create new column for difference in values using Datetime
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

# Call .describe on new column for min/max values
df['days_from_earliest_credit_to_issue'].describe()

count    128412.000000
mean       5859.891490
std        2886.535578
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

##### Utilizing .dt.days we were able to determine that the shortest credit profile in this dataset is ~3 years old.

##### The length of the average credit profile is ~16 years old.

##### The oldest credit profile is a whopping ~68.96 years old!

# BEGIN EXTRA TASKS

##### Let's take a look at our 'term' column and see what we can find.

In [27]:
df['term'].dtypes

dtype('O')

##### Looks like we're dealing with another string, let's clean this up with .strip

In [36]:
# Define function to convert string to integer
def str_to_int(cell_contents):
  return int(cell_contents.strip().strip('Months').strip(' m'))

# Call our str_to_int function on 'term'
df['term'] = df['term'].apply(str_to_int)

In [37]:
# Now we will call a specific instance of 'term' to verify conversion
df['term'][2]

60

In [38]:
# Now we will call .dtypes on 'term'
df['term'].dtypes

dtype('int64')

##### Great! Now that we have 'term' set to an Integer value we will create a new column in relation to 'loan_status'

In [39]:
# Define function to convert 'loan_status'
def is_great(cell_contents):
  a = 1
  b = 0
  if cell_contents == 'Current':
    return a
  elif cell_contents == 'Fully Paid':
    return a
  else:
    return b

# Apply function on the new column
df['loan_status_is_great'] = df['loan_status'].apply(is_great)

In [40]:
df['loan_status_is_great'].head()

0    1
1    1
2    1
3    1
4    0
Name: loan_status_is_great, dtype: int64

In [41]:
df['loan_status_is_great'].dtypes

dtype('int64')

##### Now we will create two new columns for the last payment year and month.

In [42]:
df['last_pymnt_d']

0         Jan-2020
1         Aug-2020
2         Aug-2020
3         Aug-2020
4         Sep-2019
            ...   
128407    Dec-2018
128408    Aug-2020
128409    Aug-2020
128410    Aug-2020
128411    Aug-2020
Name: last_pymnt_d, Length: 128412, dtype: object

In [43]:
# Convert 'last_pymnt_d' to Datetime
df['last_pymnt_d'] = pd.to_datetime(df['last_pymnt_d'], infer_datetime_format=True)

In [44]:
df['last_pymnt_d'] = df['last_pymnt_d'].dropna()
df['last_pymnt_d'].value_counts()

2020-08-01    81163
2020-07-01     3258
2019-07-01     3146
2019-09-01     2755
2019-10-01     2722
2019-08-01     2716
2019-12-01     2713
2019-05-01     2547
2020-06-01     2412
2019-11-01     2388
2019-04-01     2356
2020-01-01     2304
2020-03-01     2289
2019-06-01     2288
2020-05-01     1985
2019-03-01     1984
2020-04-01     1932
2019-02-01     1915
2020-02-01     1782
2019-01-01     1561
2018-12-01     1054
2018-11-01      679
2018-10-01      234
2020-09-01       70
Name: last_pymnt_d, dtype: int64

In [45]:
df['last_pymnt_d_month'] = df['last_pymnt_d'].dt.month
df['last_pymnt_d_year'] = df['last_pymnt_d'].dt.year

In [46]:
df['last_pymnt_d'].isnull().sum()

159

In [47]:
df.dtypes[-2:]

last_pymnt_d_month    float64
last_pymnt_d_year     float64
dtype: object

# Stretch Goals

You can do more with the LendingClub or Instacart datasets.

LendingClub options:
- There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.
- Modify the `emp_title` column to replace titles with 'Other' if the title is not in the top 20. 
- Take initiatve and work on your own ideas!

Instacart options:
- Read [Instacart Market Basket Analysis, Winner's Interview: 2nd place, Kazuki Onodera](http://blog.kaggle.com/2017/09/21/instacart-market-basket-analysis-winners-interview-2nd-place-kazuki-onodera/), especially the **Feature Engineering** section. (Can you choose one feature from his bulleted lists, and try to engineer it with pandas code?)
- Read and replicate parts of [Simple Exploration Notebook - Instacart](https://www.kaggle.com/sudalairajkumar/simple-exploration-notebook-instacart). (It's the Python Notebook with the most upvotes for this Kaggle competition.)
- Take initiative and work on your own ideas!

You can uncomment and run the cells below to re-download and extract the Instacart data

In [ ]:
# !wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

In [ ]:
# !tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

In [ ]:
# %cd instacart_2017_05_01